Fitbit.Portable/FitBitHttpClientMessageHandler.cs

```diff
@@ -1,38 +1,59 @@
 ﻿namespace Fitbit.Api.Portable
 {
+    using System;
     using System.Diagnostics;
     using System.Net.Http;
     using System.Threading;
     using System.Threading.Tasks;
 
-    public class FitbitHttpClientMessageHandler : DelegatingHandler
+    internal class FitbitHttpClientMessageHandler : DelegatingHandler
     {
-        private IFitbitClientInterceptor logger;
+        private IFitbitInterceptor interceptor;
+        Func<Task<HttpResponseMessage>, CancellationToken, Task<HttpResponseMessage>> responseHandler;
 
-        public FitbitHttpClientMessageHandler(IFitbitClientInterceptor logger)
+        //private ITokenManager tokenManager;
+
+        public FitbitHttpClientMessageHandler(IFitbitInterceptor interceptor)
         {
-            this.logger = logger;
+            this.interceptor = interceptor;
+            //this.tokenManager = tokenManager;
+            responseHandler = ResponseHandler; 
             //Define the inner must handler. Otherwise exception is thrown.
             InnerHandler = new HttpClientHandler();
         }
 
+        //Handle the following method with EXTREME care as it will be invoked on ALL requests made by FitbitClient
         // We override the SendAsync method to intercept both the request and response path
         protected override Task<HttpResponseMessage> SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
         {
+            Task<HttpResponseMessage> interceptorResponse = null;
             Debug.WriteLine("Entering Http client's request message handler. Request details: {0}", request.ToString());
-            if (logger != null)
-                logger.InterceptRequest(request, cancellationToken);
-
-            return base.SendAsync(request, cancellationToken).ContinueWith(
-                     requestTask =>
-                     {
-                         Debug.WriteLine("Entering Http client's response message handler. Response details: {0}", requestTask.Result);
-                         if (logger != null)
-                             logger.InterceptResponse(requestTask.Result, cancellationToken);
-
-                         return requestTask.Result;
-                     }, TaskContinuationOptions.OnlyOnRanToCompletion
+
+            if (interceptor != null)
+                interceptorResponse = interceptor.InterceptRequest(request, cancellationToken);
+
+            if(interceptorResponse != null) //then highjack the request pipeline and return the HttpResponse returned by interceptor. Invoke Response handler at return.
```

Comment:
The highjack is to simplify testing (see sample usage in [oauth2-autorefresh branch] (https://github.com/mxa0079/Fitbit.NET/blob/oauth2-autorefresh/Fitbit.Portable.Tests/FitbitClientHttpMessageHandlerTests.cs#L82) ). It addresses your accurate comments from the other day of enabling testing scenarios without having to pass an HttpClient to constructor. I think your message handler pipeline is a good idea, and it can coexist with the ability to "highjack" the request pipeline.